pkg/spdx/gomod.go

```diff
@@ -0,0 +1,297 @@
+	ScanPackageLicense(*GoPackage, *license.Reader, *GoModuleOptions) error
+}
+
+// Initializes a go module from the specified path
+func (mod *GoModule) Open() error {
+	gomod, err := mod.impl.OpenModule(mod.opts)
+	if err != nil {
+		return errors.Wrap(err, "opening module")
+	}
+	mod.GoMod = gomod
+
+	// Build the package list
+	pkgs, err := mod.impl.BuildPackageList(mod.GoMod)
+	if err != nil {
+		return errors.Wrap(err, "building module package list")
+	}
+	mod.Packages = pkgs
+	return nil
+}
+
+// RemoveDownloads cleans all downloads
+func (mod *GoModule) RemoveDownloads() error {
+	return mod.impl.RemoveDownloads(mod.Packages)
+}
+
+// DownloadPackages downloads all the module's packages to the local disk
+func (mod *GoModule) DownloadPackages() error {
+	logrus.Infof("Downloading source code for %d packages", len(mod.Packages))
+	if mod.Packages == nil {
+		return errors.New("Unable to download packages, package list is nil")
+	}
+
+	for _, pkg := range mod.Packages {
+		if err := mod.impl.DownloadPackage(pkg, mod.opts, true); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ScanLicenses scans the licenses and populats the fields
+func (mod *GoModule) ScanLicenses() error {
+	if mod.Packages == nil {
+		return errors.New("Unable to scan lincese files, package list is nil")
+	}
+
+	reader, err := mod.impl.LicenseReader()
+	if err != nil {
+		return errors.Wrap(err, "creating license scanner")
+	}
+
+	// Do a quick re-check for missing downloads
+	// todo: paralelize this. urgently.
+	for _, pkg := range mod.Packages {
+		// Call download with no force in case local data is missing
+		if err := mod.impl.DownloadPackage(pkg, mod.opts, false); err != nil {
+			// If we're unable to download the module we dont treat it as
+			// fatal, package will remain without license info but we go
+			// on scanning the rest of the packages.
+			logrus.Error(err)
+			continue
+		}
+
+		if err := mod.impl.ScanPackageLicense(pkg, reader, mod.opts); err != nil {
+			return errors.Wrapf(err, "scanning package %s for licensing info", pkg.ImportPath)
+		}
+	}
+
+	return nil
+}
+
+type GoModDefaultImpl struct {
+	licenseReader *license.Reader
+}
+
+// OpenModule opens the go,mod file for the module and parses it
+func (di *GoModDefaultImpl) OpenModule(opts *GoModuleOptions) (*modfile.File, error) {
+	modData, err := os.ReadFile(filepath.Join(opts.Path, GoModFileName))
+	if err != nil {
+		return nil, errors.Wrap(err, "reading module's go.mod file")
+	}
+	gomod, err := modfile.ParseLax("file", modData, nil)
+	if err != nil {
+		return nil, errors.Wrap(err, "reading go.mod")
+	}
+	logrus.Infof(
+		"Parsed go.mod file for %s, found %d packages",
+		gomod.Module.Mod.Path,
+		len(gomod.Require),
+	)
+	return gomod, nil
+}
+
+// BuildPackageList builds a slice of packages to assign to the module
+func (di *GoModDefaultImpl) BuildPackageList(gomod *modfile.File) ([]*GoPackage, error) {
+	pkgs := []*GoPackage{}
+	for _, req := range gomod.Require {
```

Comment:
Drive-by comment from our Twitter chat - I don't think it's actually this simple to get a full list of go dependencies. I've run into this a bunch of times, but a `go.mod` file does not actually contain every dependency used by an application, only the ones that require something other than their "natural" version as selected by the MVS algorithm.

This is actually the opposite of the mistake made in cyclone-dx I reported over here: https://github.com/CycloneDX/cyclonedx-gomod/issues/20

The only "right" way I know of is to exec "go list" directly, as they're doing in this PR: https://github.com/CycloneDX/cyclonedx-gomod/pull/22

Reply:
Thanks a lot for the feedback @dlorenc !

I thought about this problem too, but from what I concluded is that the extra set of packages you get there are dependencies of your dependencies, so in _theory at least_, listing the first level of deps should produce a valid sbom as our doc points to the next level of the dep tree. From there it would be up to each package to detail their own dependencies.

Now, as this is not happening anytime soon, it makes total sense to have a way to generate those ourselves. We could have a  `--traverse-deps` flag or something like that that grabs `go list -deps` and then does the same it is doing now but taking all versions found in `go.sum` and generating packages for each.

It will produce a huge list of spdx packages in some cases but also a complete dep tree. I'll run some experiments today.

Reply:
If I get something from you, I expect you to give me a complete list of what you're giving, direct AND transitive.  An sbom of just direct dependencies might be well formed and "valid" from a formatting/spec perspective, but from a consumer standpoint I'm not going to consider if valid if it's incomplete.  Probably not exactly relevant to your current case, but depending on the state of your build reproducibility and whether the build is hermetic, it might not be possible even for me to correctly chase down your build's transitive deps outside of your build process.  Being regularly burned by this given the state of the ecosystem today, folks are likely to be looking for as much completeness as possible.  And if burned by a missing transitive dep unmanaged and leading to a compliance issue in their environment which they attribute to your incompleteness, their trust in you might be undermined.  For this reason I'd want it defaulting to full, and at most having a flag to omit transitive or emit them separately.

I do see some possible benefit in separating direct and transitive deps in listings, depending (har har) on whether they're source code level deps or build stage deps or runtime deps.  This might allow for more efficient rebuilds when deps move forward (eg: consider the follow on consequence difference of a change in a static linked code dep vs containerized runtime dep).

Reply:
As always, thanks for your invaluable insight @tpepper ! I agree completely, that's the feeling I tried to express from reading @dlorenc 's comment. Given the state of the world, we will have to go the extra mile to produce a truly useful SBOM.  I had not thought about making it the default but I'll do that.

Yesterday I started working on a follow up change to alter the behavior as described, I'll PR it as we are done with this one which is already quite hefty.

Reply:
OK, the code to pull all transient deps is in: https://github.com/kubernetes/release/pull/2085 I put a hold on it until this one merges.

Thanks for your feedback!!